### PR TITLE
webapp, rest api compatibility changes

### PR DESCRIPTION
**Pull request by SpiReCZ**
This is PR for core compatibility changes to enable development of other Frontend implementations like Webapp or REST API. Example implementation of REST API: https://github.com/SpiReCZ/ulozto-streamer/tree/ulozto-streamer-sb

Things i've changed in uldlib:
- **.udown is needed** to read sequentially the downloaded file while downloading it and after download has finished (SegFileReader implementation).
- **Access to Tor instance is needed** for correct handling of startup/shutdown of it`s process.
- **Access to some Downloader properties** of the downloaded file needs to be available.
- Tor data dir should be configurable since it the base folder can be set to /tmp
- sys.exit(1) is not friendly to Web App, exception should be raised instead (not all instances were fixed/changed, only some that made the most problems)
- **Not every Frontend supports prompt**
- Other minor stuff
- SegFile parent class created (needed for SegFileReader)

### CODE REVIEW

**Comment by setnicka** [comment]
Sorry for the delay, I will try to review this in the following week.

**Review by setnicka — CHANGES_REQUESTED** [review]
I review the changes, it looks good :) Some minor issues in the comments.

Also please rebase on the current master and include a summary of changed things in the commit message (like "Things i've changed in uldlib:" in the pull request description). Thank you.

File: uldlib/downloader.py [hunk 1]
```diff
@@ -291,7 +307,7 @@ def download(self, url, parts=10, target_dir="", conn_timeout=DEFAULT_CONN_TIMEO
 
         if self.isLimited:
             # no need for another CAPTCHAs
-            self.stop_captcha.set()
+            # self.stop_captcha.set() # this effectively shutdowns fastapi server if triggered
```
[the remark this answers is not in the document]
Comment (SpiReCZ) [reply uldlib/downloader.py]: I think this a left over of previous version before Frontend and threading usage. I tested it without this change and now it works.

**Comment by SpiReCZ** [comment]
added fix: argument had a wrong type (left over before Frontend)

**Comment by SpiReCZ** [comment]
The changes are ready to be merged.

**Comment by setnicka** [comment]
Thank you, merging now :)

I will publish (and upload to PyPI) a new version of the ulozto-downloader with these changes later this weekend.

**Review by setnicka — APPROVED** [review]

**Comment by SpiReCZ** [comment]
> Thank you, merging now :)
> 
> I will publish (and upload to PyPI) a new version of the ulozto-downloader with these changes later this weekend.

Thx, i already decoupled my source code and it is ready to use the downloader as a dependency.

**Pull request merged**